task/host-sp-comms/src/main.rs

```diff
@@ -777,6 +784,55 @@ impl ServerImpl {
                     },
                 );
             }
+            Key::InstallinatorImageId => {
+                // Borrow `cp_agent` to avoid borrowing `self` in the closure
+                // below.
+                let cp_agent = &self.cp_agent;
+
+                // We don't want to have to set aside our own memory to copy the
+                // installinator image ID (other than our already-allocated
+                // outgoing tx buf), so we will optimistically serialize a
+                // successful response, including the image ID. After
+                // serializing this successful response, we'll check that
+                // `max_response_len` (i.e., the buffer length of the host
+                // process that requested this value) is sufficient; if it is
+                // not (or if we have no installinator image ID at all), we'll
+                // discard the optimistically-serialized response and return an
+                // error.
+                //
+                // We expect both of these "reset and replace the response with
+                // an error" to be extremely rare: host processes should not ask
+                // for an installinator ID with a too-small buffer, and should
+                // only ask for an installinator ID during a recovery process in
+                // which we expect MGS has already given us an ID.
+                let mut response_len = 0;
+                self.tx_buf.encode_response(
+                    sequence,
+                    &SpToHost::KeyLookupResult(KeyLookupResult::Ok),
+                    |mut buf| {
+                        // Statically guarantee we have sufficient space in
+                        // `buf` for the installinator image ID blob, and then
+                        // cap `buf` to that length to satisfy the idol
+                        // operation length limit.
+                        const_assert!(
+                            MIN_SP_TO_HOST_FILL_DATA_LEN
+                                >= MAX_INSTALLINATOR_IMAGE_ID_LEN
+                        );
+                        buf = &mut buf[..MAX_INSTALLINATOR_IMAGE_ID_LEN];
+
+                        response_len = cp_agent.get_installinator_image_id(buf);
+                        response_len
+                    },
+                );
+
+                if response_len == 0 {
```

Comment:
Is it possible for `control-plane-agent` to load a zero-length installinator image ID, or is that nonsensical?

Reply:
I think we want some way of unsetting the value, so the control plane can clear out the image ID after recovery completes for cleanliness. Given an image ID has to have _something_ in it, I was using "the image ID vec is empty" as the sentinel value for "we have no image ID", meaning the control plane would intentionally clear the vec out (leading to `response_len == 0` here) after recovery completes.

If this is too clever I could use `Option`s instead; `Option<Vec<u8>>` always feels weird to me though when `Some([])` is effectively the same as `None`.

Reply:
Okay, that plan seems reasonable to me!

Reply:
Added a comment in 68473bc3 explaining this conversion.